### PR TITLE
Poor video scaling quality #53080

### PR DESCRIPTION
**Pull request by zljj0818**
## Description
Use bilinear instead of nearest filter to draw surface texture

## Related Issues
https://github.com/flutter/flutter/issues/53080

## Tests

I added the following tests:

- TextureLayerTest

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x ] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x ] I signed the [CLA].
- [x ] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x ] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x ] I updated/added relevant documentation.
- [x ] All existing and new tests are passing.
- [ x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [ x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by zljj0818** [comment]
@googlebot I signed it!

**Comment by zljj0818** [comment]
@googlebot I signed it!

**Comment by zljj0818** [comment]
@googlebot I signed it!

**Comment by zljj0818** [comment]
@googlebot I signed it!

**Comment by zljj0818** [comment]
@liyuqian please review this PR

**Comment by liyuqian** [comment]
The fix looks good! My main concern is how to write a test for this (see also [this comment](https://github.com/flutter/flutter/issues/53080#issuecomment-638318765)). I felt that we need to either invest some resources to enable platform-specific golden image tests, or maybe make an exception to let this PR land without a test.

CC @zoeyfan to figure out the priority of https://github.com/flutter/flutter/issues/53080 for the customer, and see whether we should invest more resources.

File: shell/platform/android/android_external_texture_gl.cc [hunk 1]
```diff
@@ -64,7 +64,9 @@ void AndroidExternalTextureGL::Paint(SkCanvas& canvas,
       transformAroundCenter.postTranslate(0.5, 0.5);
       canvas.concat(transformAroundCenter);
     }
-    canvas.drawImage(image, 0, 0);
+    SkPaint paint;
+    paint.setFilterQuality(kLow_SkFilterQuality);
```
Comment (flar) [inline shell/platform/android/android_external_texture_gl.cc:68]: This fix is making an assumption about which filter quality is desired. Should there be a way for the app to control the filter quality through the framework?

Comment (liyuqian) [reply shell/platform/android/android_external_texture_gl.cc:68]: Yes, providing a configurable option would be better. One should be able to add that option to engine `TextureLayer` and framework `Texture` widget similar to how to we added `clipBehavior` to engine `ClipPathLayer` and framework `ClipPath` widget.

If there's no such configurable option, I felt that the bilinear `kLow_SkFilterQuality` should be preferred over the nearest neighbor `kNone_SkFilterQuality`, unless the matrix has only translations (no perspective transformation and the scale is 1). Also CC @chinmaygarde for more thoughts.

Comment (flar) [reply shell/platform/android/android_external_texture_gl.cc:68]: I agree that the default for the parameter should be FilterQuality.low.

Another point of reference for transferring enum data from the framework to the engine can be seen when the external-facing Paint object in lib/ui/painting.dart takes a FilterQuality parameter and converts it to an index which is then reconstructed into a native Skia filter value in lib/ui/painting/paint.cc

Comment (zljj0818) [reply shell/platform/android/android_external_texture_gl.cc:68]: ok, i will add it  as image's filter in this PR:https://github.com/flutter/flutter/pull/23329

Comment (zljj0818) [reply shell/platform/android/android_external_texture_gl.cc:68]: @flar @liyuqian Do i need to post another PR that include the framework change?

**Comment by zljj0818** [comment]
This is framework PR: https://github.com/flutter/flutter/pull/58958
But it failed due to miss related engine commit

**Comment by zljj0818** [comment]
+ cd ..
+ dart web_sdk/test/api_conform_test.dart
Checking 72 public classes.
Warning: lib/ui/ui.dart SceneBuilder.addTexture has a different parameter length than in lib/web_ui/ui.dart.
Failure!

Why format_and_dart_test  failed? 
I modify web_ui, but web_tests failed :(

**Comment by liyuqian** [comment]
For the `web_ui` failure, you'll need to modify https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/ui/compositing.dart

**Comment by zljj0818** [comment]
Mac build seem has some exception. 
@flar @liyuqian please take a look at current change

**Comment by flar** [comment]
The mac builds seem to have various infra issues. There are discussions in the internal roller chats about these failures already.

The new changes look fine to me, but I think it might help to have a unit test for the new texture option. It should be as simple as cut/paste/rename one of the existing simple texture unit tests and then switch the filter quality to another value.

Other than that, I'll leave it to @liyuqian or @chinmaygarde to look it over for completeness (i.e. if we're overlooking an impact this might have on other code).

We also need to be prepared for a change in a benchmark if any of the macrobenchmark apps use a Texture internally. We'll need to justify the difference and rebase them accordingly.

**Comment by zljj0818** [comment]
> The mac builds seem to have various infra issues. There are discussions in the internal roller chats about these failures already.
> 
> The new changes look fine to me, but I think it might help to have a unit test for the new texture option. It should be as simple as cut/paste/rename one of the existing simple texture unit tests and then switch the filter quality to another value.
> 
> Other than that, I'll leave it to @liyuqian or @chinmaygarde to look it over for completeness (i.e. if we're overlooking an impact this might have on other code).
> 
> We also need to be prepared for a change in a benchmark if any of the macrobenchmark apps use a Texture internally. We'll need to justify the difference and rebase them accordingly.
unittests added.

**Comment by zljj0818** [comment]
> The mac builds seem to have various infra issues. There are discussions in the internal roller chats about these failures already.
> 
> The new changes look fine to me, but I think it might help to have a unit test for the new texture option. It should be as simple as cut/paste/rename one of the existing simple texture unit tests and then switch the filter quality to another value.
> 
> Other than that, I'll leave it to @liyuqian or @chinmaygarde to look it over for completeness (i.e. if we're overlooking an impact this might have on other code).
> 
> We also need to be prepared for a change in a benchmark if any of the macrobenchmark apps use a Texture internally. We'll need to justify the difference and rebase them accordingly.

I have added unittests, please review

File: shell/platform/android/android_external_texture_gl.cc (hunk 1 above)
Comment (liyuqian) [reply shell/platform/android/android_external_texture_gl.cc]: Yes, I see you've already posted https://github.com/flutter/flutter/pull/58958

**Comment by liyuqian** [comment]
Unit tests look good! @chinmaygarde : would you be Ok with this approach of adding a `FilterQuality` option to `SceneBuilder::AddTextureLayer`?

Regarding the failed LUCI tests, they seem to be all cancelled unexpectedly for some reasons. @godofredoc @dnfield : did you see any problem there? How can we re-trigger those tests?

**Comment by godofredoc** [comment]
> Unit tests look good! @chinmaygarde : would you be Ok with this approach of adding a `FilterQuality` option to `SceneBuilder::AddTextureLayer`?
> 
> Regarding the failed LUCI tests, they seem to be all cancelled unexpectedly for some reasons. @godofredoc @dnfield : did you see any problem there? How can we re-trigger those tests?

For now the only way to re-trigger tests is by uploading a new commit. Adding functionality to re-trigger tests is work in progress: https://github.com/flutter/flutter/issues/56422

**Comment by zljj0818** [comment]
@liyuqian @flar @chinmaygarde Checkes OK, please review:)

File: shell/platform/android/android_external_texture_gl.cc (hunk 1 above)
Comment (liyuqian) [reply shell/platform/android/android_external_texture_gl.cc]: @flar  are you still requiring some changes regarding the newest patch?

**Comment by zljj0818** [comment]
@liyuqian  What about the check pendings?

**Comment by liyuqian** [comment]
The new pending tasks (Linux/Mac Engine Drone) don't even have "Details" links. I'm not sure whether they passed or not. @godofredoc : how to we check if those tasks are successful?

**Comment by christopherfujino** [comment]
Here are the builder consoles for the two pending tasks:

https://ci.chromium.org/p/flutter/builders/try/Mac%20Engine%20Drone?&limit=100
https://ci.chromium.org/p/flutter/builders/try/Linux%20Engine%20Drone?&limit=100

I also filed an issue to allow identifying which PR a try build is associated with https://github.com/flutter/flutter/issues/59581

**Comment by liyuqian** [comment]
Thanks @christopherfujino ! For now, it seems to be probably easier for @zljj0818 to just `git rebase master && git push -f` to trigger the tests again?

**Comment by godofredoc** [comment]
> Here are the builder consoles for the two pending tasks:
> 
> https://ci.chromium.org/p/flutter/builders/try/Mac%20Engine%20Drone?&limit=100
> https://ci.chromium.org/p/flutter/builders/try/Linux%20Engine%20Drone?&limit=100
> 
> I also filed an issue to allow identifying which PR a try build is associated with [flutter/flutter#59581](https://github.com/flutter/flutter/issues/59581)

x drone builders should not show up as check on the PRs as they are used for sub-builds. This is an unintended consequence of adding the checks immediately after triggering builds. I'd suggest to add https://github.com/flutter/flutter/issues/59526 as reference and submit. I'll send a PR to fix the problem during the day.

**Comment by zljj0818** [comment]
> Thanks @christopherfujino ! For now, it seems to be probably easier for @zljj0818 to just `git rebase master && git push -f` to trigger the tests again?

It's ok now :)

**Pull request merged**